### PR TITLE
`join_to_delete` is same as `join_to_update`

### PR DESCRIPTION
**Pull request by kamipo**

### CODE REVIEW

**Comment by rafaelfranca** [comment]
There are failures in the some database adapters.

File: activerecord/lib/active_record/relation.rb [hunk 1]
```diff
@@ -371,10 +371,9 @@ def update_all(updates)
 
       stmt.set Arel.sql(@klass.send(:sanitize_sql_for_assignment, updates))
       stmt.table(table)
-      stmt.key = table[primary_key]
```
Comment (kamipo) [inline activerecord/lib/active_record/relation.rb]: test failures caused by removing this line. I fix to restore this line.

**Comment by sgrif** [comment]
Is this true of SQL Server and Oracle as well? @metaskills @yahonda

**Comment by metaskills** [comment]
The SQL Server adapter has yet to worry about `join_to_delete` and/or `join_to_update`, so I doubt this affects us.

File: activerecord/lib/active_record/relation.rb [hunk 2]
```diff
@@ -374,7 +374,7 @@ def update_all(updates)
       stmt.key = table[primary_key]
 
       if joins_values.any?
-        @klass.connection.join_to_update(stmt, arel)
+        @klass.connection.join_to_update(stmt, arel, table[primary_key])
```
Comment (yui-knk) [inline activerecord/lib/active_record/relation.rb]: Which is better `table[primary_key]` or `stmt.key` as 3rd argument ?
Before in `join_to_update` `key` was `stmt.key`, so I feel `stmt.key` is better .

**Comment by rafaelfranca** [comment]
> Is this true of SQL Server and Oracle as well?

If it is not true they would already have their own implementations, so this PR should not affect them.

**Comment by rafaelfranca** [comment]
I should have searched before. So, changing `join_to_update` to accept three arguments will break [oracle-enhanced](https://github.com/rsim/oracle-enhanced/blob/3c42131db82b64ac41645db3affc6e4650289df6/lib/active_record/connection_adapters/oracle_enhanced_adapter.rb#L1254), so even that this refactoring is good we can't do without a proper deprecation plan.

**Comment by rafaelfranca** [comment]
Reverted in 5082cce.

**Comment by rafaelfranca** [comment]
Just to be clear, not saying that this change is not going to be accepted, just trying to play safe. @yahonda do you think it is fine to change oracle-enhanced to accept the third argument?

@kamipo can you open a new PR with the same change, if not I can apply the changes again later.

**Comment by yahonda** [comment]
My laptop are being offline in couple of days. will have a look at this on coming Saturday. Thanks

**Comment by yahonda** [comment]
Thanks for considering about 3rd party adapter.

This commit changes api but this change is relatively small. then as long as this change applies to master branch and not backported to 4-2-stable, it is fine for me.

Each Oracle enhanced adapter minor version only supports one Rails minor version,  i.e. Oracle enhanced adapter 1.6.x supports Rails 4.2.x, then Rails 5 support will be available as Oracle enhanced adapter 1.7.x.

**Comment by yahonda** [comment]
Updated #22620

**Pull request merged**